net/network/netem-control/files/sbin/netem-control.lua

```diff
@@ -21,29 +23,38 @@ function setRules (operation)
     local qdisc_parent = 0
     local errors = 0
 
+    if ( operation == "del") then
+        --remove all rules associated to previously added interfaces
+        for k,v in ipairs(intfcfg) do
+            tc_str = "tc qdisc del dev "..v.." root"
+            exec(tc_str)
+        end
+        table.remove(intfcfg)
+        return
+    end
+
     x:foreach("netem", "interface",
     function (section)
+
         iface = x:get("netem", section[".name"], "ifname")
         if iface == nil then
             return
         end
+
         enabled = x:get("netem", section[".name"], "enabled")
         if enabled == "1" then
             qdisc_parent = qdisc_parent + 1
+            --keep track of all interfaces where we are adding rules
+            table.insert(intfcfg, iface)
 
-            if (operation == "add" or operation == "del") then
-                tc_str = "tc qdisc del dev "..iface.." root"
-                exec(tc_str)
+            if (operation == "add") then
+                --tc_str = "tc qdisc del dev "..iface.." root"
```

Comment:
I _think_ commenting out this `qdisc del dev` should be ok, but I'm pretty sure we had it delete and then replace because of issues once upon a time with trying to change a qdisc.

Reply:
Let me look into this again. What I saw was that this caused rules not to be persistent across device reboot. So the netem config would be present but the qdisc rules would be missing.

Reply:
Quickly ran through the following test cases with the code as is.

1. Added new rule for wan emulation on eth0, validated tc qdisc rules exist, test wan emulation and results are as expected
2. Reboot device, validated tc qdisc rules exist, test wan emulation and results are as expected
3. Disable emulation rule, validated tc qdisc entry is removed, test wan emulation and results are as expected
4. Re-enabled emulation rule, validated tc qdisc entry is added, test wan emulation and results are as expected
5. Deleted emulation rule, validated tc qdisc entry is removed, test wan emulation and results are as expected

Is there a condition I did not test which should be considered?

Here is some more info on the device I am testing on
```
root@OpenWrt:~# uname -an
Linux OpenWrt 5.10.146 #0 SMP Fri Oct 14 22:44:41 2022 x86_64 GNU/Linux
```

With netem-control and luci-app-netem installed this is default tc qdisc output with no rules applied.
```
BusyBox v1.35.0 (2022-10-14 22:44:41 UTC) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt 22.03.2, r19803-9a599fee93
 -----------------------------------------------------
root@OpenWrt:~# tc qdisc
qdisc noqueue 0: dev lo root refcnt 2
qdisc mq 0: dev eth0 root
qdisc fq_codel 0: dev eth0 parent :4 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64
qdisc fq_codel 0: dev eth0 parent :3 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64
qdisc fq_codel 0: dev eth0 parent :2 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64
qdisc fq_codel 0: dev eth0 parent :1 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64
qdisc mq 0: dev eth1 root
qdisc fq_codel 0: dev eth1 parent :4 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64
qdisc fq_codel 0: dev eth1 parent :3 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64
qdisc fq_codel 0: dev eth1 parent :2 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64
qdisc fq_codel 0: dev eth1 parent :1 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64
qdisc noqueue 0: dev br-lan root refcnt 2
```

Installed packages
```
root@OpenWrt:~# opkg list-installed
base-files - 1494-r19803-9a599fee93
bnx2-firmware - 20220411-1
busybox - 1.35.0-4
ca-bundle - 20211016-1
cgi-io - 2022-08-10-901b0f04-21
dnsmasq - 2.86-14
dropbear - 2022.82-2
e2fsprogs - 1.46.5-1
ethtool - 5.16-1
firewall4 - 2022-10-14-4fbf6d75-1
fstools - 2022-06-02-93369be0-2
fwtool - 2019-11-12-8f7fe925-1
getrandom - 2021-08-03-205defb5-2
grub2 - 2.06-3
grub2-bios-setup - 2.06-3
grub2-efi - 2.06-3
iperf3 - 3.11-1
jansson4 - 2.13.1-2
jshn - 2022-05-15-d2223ef9-1
jsonfilter - 2018-02-04-c7e938d6-1
kernel - 5.10.146-1-5d4fa2bd675198961db287e63a2370e2
kmod-amazon-ena - 5.10.146-1
kmod-amd-xgbe - 5.10.146-1
kmod-bnx2 - 5.10.146-1
kmod-button-hotplug - 5.10.146-3
kmod-crypto-crc32c - 5.10.146-1
kmod-crypto-hash - 5.10.146-1
kmod-e1000 - 5.10.146-1
kmod-e1000e - 5.10.146-1
kmod-forcedeth - 5.10.146-1
kmod-fs-vfat - 5.10.146-1
kmod-hwmon-core - 5.10.146-1
kmod-i2c-algo-bit - 5.10.146-1
kmod-i2c-core - 5.10.146-1
kmod-igb - 5.10.146-1
kmod-igc - 5.10.146-1
kmod-input-core - 5.10.146-1
kmod-ipt-core - 5.10.146-1
kmod-ixgbe - 5.10.146-1
kmod-lib-crc-ccitt - 5.10.146-1
kmod-lib-crc32c - 5.10.146-1
kmod-lib-textsearch - 5.10.146-1
kmod-libphy - 5.10.146-1
kmod-mdio - 5.10.146-1
kmod-mdio-devres - 5.10.146-1
kmod-mii - 5.10.146-1
kmod-netem - 5.10.146-1
kmod-nf-conntrack - 5.10.146-1
kmod-nf-conntrack6 - 5.10.146-1
kmod-nf-flow - 5.10.146-1
kmod-nf-ipt - 5.10.146-1
kmod-nf-log - 5.10.146-1
kmod-nf-log6 - 5.10.146-1
kmod-nf-nat - 5.10.146-1
kmod-nf-reject - 5.10.146-1
kmod-nf-reject6 - 5.10.146-1
kmod-nfnetlink - 5.10.146-1
kmod-nft-core - 5.10.146-1
kmod-nft-fib - 5.10.146-1
kmod-nft-nat - 5.10.146-1
kmod-nft-offload - 5.10.146-1
kmod-nls-base - 5.10.146-1
kmod-nls-cp437 - 5.10.146-1
kmod-nls-iso8859-1 - 5.10.146-1
kmod-nls-utf8 - 5.10.146-1
kmod-phy-realtek - 5.10.146-1
kmod-ppp - 5.10.146-1
kmod-pppoe - 5.10.146-1
kmod-pppox - 5.10.146-1
kmod-pps - 5.10.146-1
kmod-ptp - 5.10.146-1
kmod-r8169 - 5.10.146-1
kmod-sched - 5.10.146-1
kmod-sched-core - 5.10.146-1
kmod-slhc - 5.10.146-1
kmod-tg3 - 5.10.146-1
libblkid1 - 2.37.3-3
libblobmsg-json20220515 - 2022-05-15-d2223ef9-1
libc - 1.2.3-4
libcomerr0 - 1.46.5-1
libext2fs2 - 1.46.5-1
libf2fs6 - 1.14.0-3
libgcc1 - 11.2.0-4
libiwinfo-data - 2022-08-19-0dad3e66-1
libiwinfo-lua - 2022-08-19-0dad3e66-1
libiwinfo20210430 - 2022-08-19-0dad3e66-1
libjson-c5 - 0.15-2
libjson-script20220515 - 2022-05-15-d2223ef9-1
liblua5.1.5 - 5.1.5-10
liblucihttp-lua - 2022-07-08-6e68a106-1
liblucihttp0 - 2022-07-08-6e68a106-1
libmnl0 - 1.0.5-1
libnftnl11 - 1.2.1-1
libnl-tiny1 - 2021-11-21-8e0555fb-1
libpthread - 1.2.3-4
librt - 1.2.3-4
libsmartcols1 - 2.37.3-3
libss2 - 1.46.5-1
libubox20220515 - 2022-05-15-d2223ef9-1
libubus-lua - 2022-06-01-2bebf93c-1
libubus20220601 - 2022-06-01-2bebf93c-1
libuci-lua - 2021-10-22-f84f49f0-6
libuci20130104 - 2021-10-22-f84f49f0-6
libuclient20201210 - 2021-05-14-6a6011df-1
libucode20220812 - 2022-10-07-4ae70721-1
libustream-wolfssl20201210 - 2022-01-16-868fd881-2
libuuid1 - 2.37.3-3
libwolfssl5.5.1.ee39414e - 5.5.1-stable-3
logd - 2021-08-03-205defb5-2
lua - 5.1.5-10
lua-bit32 - 5.3.0-1
luaposix - 35.1-1
luci - git-20.074.84698-ead5e81
luci-app-firewall - git-22.089.67563-7e3c1b4
luci-app-netem - 0.1-1
luci-app-opkg - git-22.273.29004-9f6876b
luci-base - git-22.288.45147-96ec0cd
luci-compat - git-22.069.45071-03bb0e2
luci-lib-base - git-20.232.39649-1f6dc29
luci-lib-ip - git-20.250.76529-62505bd
luci-lib-jsonc - git-22.097.61921-7513345
luci-lib-nixio - git-20.234.06894-c4a4e43
luci-mod-admin-full - git-19.253.48496-3f93650
luci-mod-network - git-22.244.54818-b13d8c7
luci-mod-status - git-22.189.48501-6731190
luci-mod-system - git-22.264.46172-b6b7da4
luci-proto-ipv6 - git-21.148.48881-79947af
luci-proto-ppp - git-21.158.38888-88b9d84
luci-ssl - git-20.244.36115-e10f954
luci-theme-bootstrap - git-22.288.45147-96ec0cd
mkf2fs - 1.14.0-3
mtd - 26
netem-control - 0.2-1
netifd - 2022-08-25-76d2d41b-1
nftables-json - 1.0.2-2.1
odhcp6c - 2022-08-05-7d21e8d8-18
odhcpd-ipv6only - 2022-03-22-860ca900-1
openwrt-keyring - 2022-03-25-62471e69-3
opkg - 2022-02-24-d038e5b6-1
partx-utils - 2.37.3-3
ppp - 2.4.9.git-2021-01-04-3
ppp-mod-pppoe - 2.4.9.git-2021-01-04-3
procd - 2022-06-01-7a009685-1
procd-seccomp - 2022-06-01-7a009685-1
procd-ujail - 2022-06-01-7a009685-1
px5g-wolfssl - 5.1
r8169-firmware - 20220411-1
rpcd - 2022-09-21-8c852b65-1
rpcd-mod-file - 2022-09-21-8c852b65-1
rpcd-mod-iwinfo - 2022-09-21-8c852b65-1
rpcd-mod-luci - 20210614
rpcd-mod-rrdns - 20170710
tc-tiny - 5.15.0-3
ubox - 2021-08-03-205defb5-2
ubus - 2022-06-01-2bebf93c-1
ubusd - 2022-06-01-2bebf93c-1
uci - 2021-10-22-f84f49f0-6
uclient-fetch - 2021-05-14-6a6011df-1
ucode - 2022-10-07-4ae70721-1
ucode-mod-fs - 2022-10-07-4ae70721-1
ucode-mod-ubus - 2022-10-07-4ae70721-1
ucode-mod-uci - 2022-10-07-4ae70721-1
uhttpd - 2022-08-12-e3395cd9-1
uhttpd-mod-ubus - 2022-08-12-e3395cd9-1
urandom-seed - 3
urngd - 2020-01-21-c7f7b6b6-1
usign - 2020-05-23-f1f65026-1
```